### ISSUE TITLE
Should increasing the number of trials also increase memory usage?

### ISSUE THREAD

**Issue by ktattan** [optuna/optuna#752]
I can't see why it should, but for my application it is.

I have a 24 core CPU, and so I am setting `n_jobs=-1` to use all 24. Optimization works fine with `n_trials=10` - memory usage is about 50% on 32 Gb of RAM (so 16Gb is being used). However, if I try to increase `n_trials` to anything greater than 10, memory usage increases to 100% and the kernel kills all processes. I can't understand why this would be so.

Is there a correct way / combination to use `n_jobs` and `n_trials` - maybe `n_trials` should be an integer multiple of the number of CPU's being used? I tried a few different combinations but nothing seemed to work - memory usage always increased to 100% and the process was killed. Ideally I would like to use all 24 CPU's to get more trials done in parallel.

p.s. I don't know if it is relevant, but I saw the note on [avoiding sqlite](https://optuna.readthedocs.io/en/stable/tutorial/distributed.html) for large scale distributed optimization. I am using sqlite database to store optimization results, but I wouldn't call the optimization "large scale"; my raw data is only 175Mb and the machine learning function I am optimizing uses under 1Gb of memory when running on a single CPU (it also shares some memory when more than 1 CPU is being used).

Any help or advice is welcomed. Thank you

**Comment by ktattan** [comment]
Apologies - I just found #497 which seems to address this issue.

**Comment by sile** [comment]
Hi @ktattan,

Thank you for reporting your problem.
I think that Optuna itself doesn't require huge memory for parallel optimization even if the number of trials increases. So, this is unexpected behavior.
If you can share an example script to reproduce this problem, I would like to investigate further.

Anyway, feel free to reopen this issue if #497 cannot resolve your problem.

**Comment by ktattan** [comment]
Thank you @sile 

I have changed my code to be as close as possible to the structure in #497 and [this example](https://github.com/optuna/optuna/blob/master/examples/pruning/keras_integration.py) to try and mitigate the problems I am facing. However, I am still facing issues

```
def create_model(trial)
    arg = trial.suggest_uniform("arg", 0.5, 0.7)
    model = init_model(arg)
    return model

def objective(trial, train_data, test_data):
    model = create_model(trial)
    model.learn(train_data)
    cost = model.predict(test_data)
    return cost

def main(train_data, test_data, n_trials):
    study = optuna.load_study(study_name='my_study' storage=params_db_file)
    study.optimize(objective(train_data, test_data), n_trial=n_trials)


if __name__ == "__main__":
    import pandas as pd
    from joblib import Parallel, delayed

    n_trials=100
    train_data = pd.read_csv("my_train_data.csv")
    test_data = pd.read_csv("my_test_data.csv")

    Parallel(n_jobs=-1)([delayed(main)(train_data, test_data, n_trials) for i in range(n_trials)])
```

The issue is I don't know how to pass `trial` into `objective` when using `study.optimize(objective(train_data, test_data), n_trial=n_trials)`

**Comment by sile** [comment]
> How should I change the code to pass a `trial` object into `study.optimize`?

I think you can use `lambda` as follows: 
```Python
    study.optimize(lambda trial: objective(trial, train_data, test_data), n_trial=n_trials)
```
C.f. ["How to define objective functions that have own arguments?"](https://optuna.readthedocs.io/en/stable/faq.html#how-to-define-objective-functions-that-have-own-arguments)

**Comment by sile** [comment]
Since there are no activities for a long time on this issue, please let me close it. Feel free to reopen this or create another issue if need.
